kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -141,31 +154,56 @@ private ResultOrError handleNext(T result, @Nullable Throwable error) {
   private static class InputStreamingResponse<T> extends StreamingResponse<T> {
 
     private final MappingIterator<T> mappingIteratorInput;
+    private final StreamingResponseIdleTimeLimiter idleTimeLimiter;
+
+    private Throwable hasNextError = null;
+    private boolean closed = false;
 
     private InputStreamingResponse(
-        MappingIterator<T> mappingIteratorInput, ChunkedOutputFactory chunkedOutputFactory) {
+        MappingIterator<T> mappingIteratorInput,
+        ChunkedOutputFactory chunkedOutputFactory,
+        StreamingResponseIdleTimeLimiter idleTimeLimiter) {
       super(chunkedOutputFactory);
       this.mappingIteratorInput = requireNonNull(mappingIteratorInput);
+      this.idleTimeLimiter = requireNonNull(idleTimeLimiter);
     }
 
     public void close() {
-      try {
-        mappingIteratorInput.close();
-      } catch (IOException e) {
-        log.error("Error when closing the request stream.", e);
-      }
+      closed = true;
     }
 
     @Override
     public boolean hasNext() {
-      return mappingIteratorInput.hasNext();
+      if (closed) {
+        return false;
+      }
+      try {
+        return idleTimeLimiter.callWithTimeout(mappingIteratorInput::hasNextValue);
+      } catch (ExecutionException | UncheckedExecutionException e) {
+        hasNextError = e.getCause();
+        return true;
+      } catch (InterruptedException | MaxIdleTimeExceededException e) {
+        hasNextError = e;
+        return true;
+      }
     }
 
     @Override
     public CompletableFuture<T> next() {
+      if (closed) {
+        throw new NoSuchElementException();
+      }
```

Comment:
How is this handled in our exception mapping? Does it result in HTTP 500?

Reply:
It should. Calling `next()` when `hasNext()` is false is a programming error.